File: 백준/Silver/1711. 직각삼각형/직각삼각형.cc
#include <cstdio>

using namespace std;

constexpr int MAX_N = 1500;
constexpr int MAX_DEPTH = 3;

struct Coordinates {
    long long int x;
    long long int y;
};

int answer;
int N;
Coordinates tmp[MAX_DEPTH];
Coordinates C[MAX_N];

void go(int start, int depth);

int main() {
    /*
     * Input.
     */
    scanf("%d", &N);

    for (int i = 0; N > i; ++i) {
        scanf("%lld%lld", &C[i].x, &C[i].y);
    }

    /*
     * Compute.
     */
    go(0, 0);

    /*
     * Output.
     */
    printf("%d", answer);

    return 0;
}

void go(int start, int depth) {
    if (MAX_DEPTH == depth) {
        long long int ip0 = (tmp[1].x - tmp[0].x) * (tmp[2].x - tmp[0].x) + (tmp[1].y - tmp[0].y) * (tmp[2].y - tmp[0].y);
        long long int ip1 = (tmp[0].x - tmp[1].x) * (tmp[2].x - tmp[1].x) + (tmp[0].y - tmp[1].y) * (tmp[2].y - tmp[1].y);
        long long int ip2 = (tmp[0].x - tmp[2].x) * (tmp[1].x - tmp[2].x) + (tmp[0].y - tmp[2].y) * (tmp[1].y - tmp[2].y);

        if (0 == ip0 || 0 == ip1 || 0 == ip2) {
            ++answer;
        }

        return;
    }

    for (int i = start, endi = N - MAX_DEPTH + depth; endi >= i; ++i) {
        tmp[depth] = C[i];

        go(i + 1, depth + 1);
    }

    return;
}

File: 백준/Silver/1711. 직각삼각형/직각삼각형.py
import sys

input = sys.stdin.readline

# 점의 개수를 입력받는다.
N = int(input())

# 점의 좌표를 입력받는다.
C = [list(map(int, input().split())) for _ in range(N)]

# 정답(i.e., 직각삼각형의 개수)를 초기화한다.
answer = 0

# (모든 경우의 수를 탐색하면서) 직각삼각형의 성질을 활용하여 정답을 갱신한다.
for t1 in range(N - 2):
    for t2 in range(t1 + 1, N - 1):
        for t3 in range(t2 + 1, N):
            # [옵션1] 피타고리스 정리
            h1 = (C[t1][0] - C[t2][0]) ** 2 + (C[t1][1] - C[t2][1]) ** 2
            h2 = (C[t2][0] - C[t3][0]) ** 2 + (C[t2][1] - C[t3][1]) ** 2
            h3 = (C[t3][0] - C[t1][0]) ** 2 + (C[t3][1] - C[t1][1]) ** 2
            
            # if h1 + h2 + h3 == 2 * max(h1, h2, h3):
            #    answer += 1
            if h1 == h2 + h3 or h2 == h1 + h3 or h3 == h1 + h2:
               answer += 1

            # [옵션2] 백터의 내적
            # ip1 = (C[t2][0] - C[t1][0]) * (C[t3][0] - C[t1][0]) + (C[t2][1] - C[t1][1]) * (C[t3][1] - C[t1][1])
            # ip2 = (C[t1][0] - C[t2][0]) * (C[t3][0] - C[t2][0]) + (C[t1][1] - C[t2][1]) * (C[t3][1] - C[t2][1])
            # ip3 = (C[t1][0] - C[t3][0]) * (C[t2][0] - C[t3][0]) + (C[t1][1] - C[t3][1]) * (C[t1][1] - C[t3][1])

            # if not ip1 or not ip2 or not ip3:
            #    answer += 1

# 정답을 출력한다.
print(answer)
